### PR TITLE
Get a FileEntry for <new> imports

### PR DESCRIPTION
**Pull request by storoj**
Use preprocessor to lookup the `FileEntry` for `<new>` in the corner case that reports the "placement new" symbol usage.

This is an experimental change and I am looking for your feedback @kimgr.

I find the current implementation of the final steps of IWYU a bit overcomplicated. I mean the steps that actually find the intersection and the difference of two sets: existing imports and "desired" imports.

As far as I can see now IWYU works with "lines" (`OneIncludeOrForwardDeclareLine`), and IMO we can treat these as "strings".

I find it a bit inefficient and inaccurate, and my proposal is to use `FileEntries` for that. For every `IwyuFileInfo` we could collect the set of `FileEntries` being `#included`, as well as all the `SourceLocations` of these includes.

For example `IwyuFileInfo` could contain a map like:
```
std::map<const FileEntry*, std::set<SourceLocation>>
```
where the key is an import, and the value is all the locations of that import in the file relevant to that `IwuiFileInfo`. Then when we report a Full Use or a Forward Declaration we could also get the `FileEntry` for the corresponding symbol. In the end of AST traversal there will be two sets of `FileEntry` values and it should be much easier to calculate the deletions and the insertions.

Header mapping might be a good question, but as far as I can understand, mappings should only be applied to insertions (because IMO the existing imports that were used should not be modified - ?), and that operation could probably be made in the very end of the imports diffing algorithm.

I find the idea of migrating the logic to FileEntries very promising, and that pull request makes a first small step towards it. I am looking forward to get rid of as many of string-based operations as possible. One of the blockers was the `<new>` special case, and that change finds the FileEntry for `<new>` header.

What do you think about moving to FileEntry-based solution?

### CODE REVIEW

**Comment by kimgr** [comment]
Very cool! I'll try to look into this in more detail shortly. It would be nice if the commit message had more rationale as to why this is a valuable change. I no longer remember why I wanted to do this, but it's been mulling in the back of my head for ages without coming to anything, so if something was driving you to do this, the commit message is the right place to expand on it.

File: iwyu.cc [hunk 1]
```diff
@@ -2426,9 +2427,16 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         // parse it to '<new>' before using, so any path that does
         // that, and is clearly a c++ path, is fine; its exact
         // contents don't matter that much.
+
```
Comment (kimgr) [inline iwyu.cc:2430]: Unrelated, unnecessary whitespace change

File: iwyu.cc [hunk 2]
```diff
@@ -2426,9 +2427,16 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
...
         const FileEntry* use_file = CurrentFileEntry();
-        preprocessor_info().FileInfoFor(use_file)->ReportFullSymbolUse(
-            CurrentLoc(), "<new>", "operator new");
+        const clang::DirectoryLookup* TmpDir = nullptr;
```
Comment (kimgr) [inline iwyu.cc:2432]: We use `lowercase_with_underscores` naming for variables in IWYU, not LLVM's `PascalCase`.

This can probably be called `curdir` to signify what role it plays in the argument list.

File: iwyu.cc [hunk 3]
```diff
@@ -2426,9 +2427,16 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
...
+        auto file = compiler()->getPreprocessor().LookupFile(
```
Comment (kimgr) [inline iwyu.cc:2433]: Don't use `auto` unless the type is clearly visible on the line already. `Optional<FileEntryRef>` provides relevant information here.

File: iwyu_output.cc [hunk 4]
```diff
@@ -275,6 +275,22 @@ OneUse::OneUse(const string& symbol_name, const FileEntry* dfn_file,
     suggested_header_ = decl_filepath_;
 }
 
+OneUse::OneUse(const FileEntry* dfn_file, const string& dfn_filepath)
```
Comment (kimgr) [inline iwyu_output.cc:278]: I'm a little confused why this new overload is necessary. I played around with it on my local machine, but I didn't get my head around it.

Are there two changes in this commit -- one to use `FileEntry` and one to simplify construction at https://github.com/include-what-you-use/include-what-you-use/pull/911/files#diff-96c9906a158b909fda0f001cb5bd04615d46704df24173306894f3d228abbed1R669?

This one looks kind-of redundant, but I'm not entirely sure what's going on. One thing at a time is my motto :-)

File: iwyu_output.cc [hunk 5]
```diff
@@ -298,8 +314,8 @@ void OneUse::SetPublicHeaders() {
   const IncludePicker& picker = GlobalIncludePicker();   // short alias
   const string use_path = GetFilePath(use_loc_);
   // If the symbol has a special mapping, use it, otherwise map its file.
-  public_headers_ = picker.GetCandidateHeadersForSymbolUsedFrom(
-      symbol_name_, use_path);
+  public_headers_ =
```
Comment (kimgr) [inline iwyu_output.cc:317]: Unrelated reformatting, please revert.

**Review by kimgr — CHANGES_REQUESTED** [review]
Very nice. Hopefully we can use this as a first step to transition wholesale to `FileEntry*`. Comments inline.

File: iwyu.cc (hunk 1 above)
Comment (storoj) [reply iwyu.cc]: reverted

File: iwyu.cc (hunk 2 above)
Comment (storoj) [reply iwyu.cc]: used `curdir` as a variable name

File: iwyu.cc (hunk 3 above)
Comment (storoj) [reply iwyu.cc]: replaced `auto` with the explicit type

Comment (storoj) [reply iwyu.cc]: replaced `auto` with the explicit type

File: iwyu_output.cc (hunk 4 above)
Comment (storoj) [reply iwyu_output.cc]: Thanks for pointing that out, I actually had mixed feelings about adding another overload with less arguments. Reverted the change and reused the existing ctor.

**Review by kimgr — APPROVED** [review]
